File: Examples/Objective-C/Examples/AccessoryViews/AccessoryViewFormViewController.m
//
//  AccessoryViewFormViewController.m
//  XLForm ( https://github.com/xmartlabs/XLForm )
//
//  Copyright (c) 2015 Xmartlabs ( http://xmartlabs.com )
//
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "XLForm.h"
#import "AccessoryViewFormViewController.h"

//This macro defines if we use predicates to hide rows or do it manually the old way.
//Just comment out if you want it to run without predicates.
#define USE_PREDICATES_FOR_HIDING

@interface AccessoryViewFormViewController ()

@end

@implementation AccessoryViewFormViewController
{
#ifndef USE_PREDICATES_FOR_HIDING
    XLFormRowDescriptor * _rowShowAccessoryView;
    XLFormRowDescriptor * _rowStopDisableRow;
    XLFormRowDescriptor * _rowStopInlineRow;
    XLFormRowDescriptor * _rowSkipCanNotBecomeFirstResponderRow;
#endif
}


NSString * kAccessoryViewRowNavigationEnabled               = @"kRowNavigationEnabled";
NSString * kAccessoryViewRowNavigationShowAccessoryView     = @"kRowNavigationShowAccessoryView";
NSString * kAccessoryViewRowNavigationStopDisableRow        = @"rowNavigationStopDisableRow";
NSString * kAccessoryViewRowNavigationSkipCanNotBecomeFirstResponderRow = @"rowNavigationSkipCanNotBecomeFirstResponderRow";
NSString * kAccessoryViewRowNavigationStopInlineRow = @"rowNavigationStopInlineRow";
NSString * kAccessoryViewName = @"name";
NSString * kAccessoryViewEmail = @"email";
NSString * kAccessoryViewTwitter = @"twitter";
NSString * kAccessoryViewUrl = @"url";
NSString * kAccessoryViewDate = @"date";
NSString * kAccessoryViewTextView = @"textView";
NSString * kAccessoryViewCheck = @"check";
NSString * kAccessoryViewNotes = @"notes";




-(id)init
{
    self = [super init];
    if (self) {
        [self initializeForm];
    }
    return self;
}

-(void)initializeForm
{
    XLFormDescriptor * formDescriptor = [XLFormDescriptor formDescriptorWithTitle:@"Accessory View"];
    formDescriptor.rowNavigationOptions = XLFormRowNavigationOptionEnabled;
    XLFormSectionDescriptor * section;
    XLFormRowDescriptor * row;
    XLFormRowDescriptor * switchRow;
    
    // Configuration section
    section = [XLFormSectionDescriptor formSectionWithTitle:@"Row Navigation Settings"];
    section.footerTitle = @"Changing the Settings values you will navigate differently";
    [formDescriptor addFormSection:section];
    
    // RowNavigationEnabled
    switchRow = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewRowNavigationEnabled rowType:XLFormRowDescriptorTypeBooleanSwitch title:@"Row Navigation Enabled?"];
    switchRow.value = @YES;
    [section addFormRow:switchRow];
    
    // RowNavigationShowAccessoryView
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewRowNavigationShowAccessoryView rowType:XLFormRowDescriptorTypeBooleanCheck title:@"Show input accessory row?"];
    row.value = @((formDescriptor.rowNavigationOptions & XLFormRowNavigationOptionEnabled) == XLFormRowNavigationOptionEnabled);
    [section addFormRow:row];
#ifdef USE_PREDICATES_FOR_HIDING
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
#else
    _rowShowAccessoryView = row;
#endif
    
    // RowNavigationStopDisableRow
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewRowNavigationStopDisableRow rowType:XLFormRowDescriptorTypeBooleanCheck title:@"Stop when reach disabled row?"];
    row.value = @((formDescriptor.rowNavigationOptions & XLFormRowNavigationOptionStopDisableRow) == XLFormRowNavigationOptionStopDisableRow);
    [section addFormRow:row];
#ifdef USE_PREDICATES_FOR_HIDING
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
#else
    _rowStopDisableRow = row;
#endif
    
    // RowNavigationStopInlineRow
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewRowNavigationStopInlineRow rowType:XLFormRowDescriptorTypeBooleanCheck title:@"Stop when reach inline row?"];
    row.value = @((formDescriptor.rowNavigationOptions & XLFormRowNavigationOptionStopInlineRow) == XLFormRowNavigationOptionStopInlineRow);
    [section addFormRow:row];
#ifdef USE_PREDICATES_FOR_HIDING
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
#else
    _rowStopInlineRow = row;
#endif
    
    // RowNavigationSkipCanNotBecomeFirstResponderRow
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewRowNavigationSkipCanNotBecomeFirstResponderRow rowType:XLFormRowDescriptorTypeBooleanCheck title:@"Skip Can Not Become First Responder Row?"];
    row.value = @((formDescriptor.rowNavigationOptions & XLFormRowNavigationOptionSkipCanNotBecomeFirstResponderRow) == XLFormRowNavigationOptionSkipCanNotBecomeFirstResponderRow);
    [section addFormRow:row];
#ifdef USE_PREDICATES_FOR_HIDING
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
#else
    _rowSkipCanNotBecomeFirstResponderRow = row;
#endif
    

    
    // Basic Information - Section
    section = [XLFormSectionDescriptor formSection];
    [formDescriptor addFormSection:section];
    
    // Name
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewName rowType:XLFormRowDescriptorTypeText title:@"Name"];
    row.required = YES;
    [section addFormRow:row];
    
    // Email
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewEmail rowType:XLFormRowDescriptorTypeEmail title:@"Email"];
    // validate the email
    [row addValidator:[XLFormValidator emailValidator]];
    [section addFormRow:row];
    
    // Twitter
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewTwitter rowType:XLFormRowDescriptorTypeTwitter title:@"Twitter"];
    row.disabled = @YES;
    row.value = @"@no_editable";
    [section addFormRow:row];
    
    // Url
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewUrl rowType:XLFormRowDescriptorTypeURL title:@"Url"];
    [section addFormRow:row];
    
    // Url
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewDate rowType:XLFormRowDescriptorTypeDateInline title:@"Date Inline"];
    row.value = [NSDate new];
    [section addFormRow:row];
    
    
    section = [XLFormSectionDescriptor formSection];
    [formDescriptor addFormSection:section];
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewTextView rowType:XLFormRowDescriptorTypeTextView];
    [row.cellConfigAtConfigure setObject:@"TEXT VIEW EXAMPLE" forKey:@"textView.placeholder"];
    [section addFormRow:row];
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewCheck rowType:XLFormRowDescriptorTypeBooleanCheck title:@"Check"];
    [section addFormRow:row];
    
    section = [XLFormSectionDescriptor formSection];
    [formDescriptor addFormSection:section];
    row = [XLFormRowDescriptor formRowDescriptorWithTag:kAccessoryViewNotes rowType:XLFormRowDescriptorTypeTextView title:@"Notes"];
    [section addFormRow:row];
    
    self.form = formDescriptor;
}



#pragma mark - XLFormDescriptorDelegate

-(void)formRowDescriptorValueHasChanged:(XLFormRowDescriptor *)rowDescriptor oldValue:(id)oldValue newValue:(id)newValue
{
    [super formRowDescriptorValueHasChanged:rowDescriptor oldValue:oldValue newValue:newValue];
#ifndef USE_PREDICATES_FOR_HIDING
    NSString * kRowNavigationEnabled        = @"kRowNavigationEnabled";
    if ([rowDescriptor.tag isEqualToString:kRowNavigationEnabled]){
        if ([[rowDescriptor.value valueData] isEqual:@NO]){
            self.form.rowNavigationOptions = XLFormRowNavigationOptionNone;
            [self.form removeFormRowWithTag:kAccessoryViewRowNavigationShowAccessoryView];
            [self.form removeFormRowWithTag:kAccessoryViewRowNavigationStopDisableRow];
            [self.form removeFormRowWithTag:kAccessoryViewRowNavigationStopInlineRow];
            [self.form removeFormRowWithTag:kAccessoryViewRowNavigationSkipCanNotBecomeFirstResponderRow];
        }
        else{
            self.form.rowNavigationOptions = XLFormRowNavigationOptionEnabled;
            _rowShowAccessoryView.value = @YES;
            _rowStopDisableRow.value = @NO;
            _rowStopInlineRow.value = @NO;
            _rowSkipCanNotBecomeFirstResponderRow.value = @NO;
            [self.form addFormRow:_rowShowAccessoryView afterRow:rowDescriptor];
            [self.form addFormRow:_rowStopDisableRow afterRow:_rowShowAccessoryView];
            [self.form addFormRow:_rowStopInlineRow afterRow:_rowStopDisableRow];
            [self.form addFormRow:_rowSkipCanNotBecomeFirstResponderRow afterRow:_rowStopInlineRow];
            
        }
    }
    else
#endif
    if ([rowDescriptor.tag isEqualToString:kAccessoryViewRowNavigationStopDisableRow]){
        if ([[rowDescriptor.value valueData] isEqual:@(YES)]){
            self.form.rowNavigationOptions = self.form.rowNavigationOptions | XLFormRowNavigationOptionStopDisableRow;
        }
        else{
            self.form.rowNavigationOptions = self.form.rowNavigationOptions & (~XLFormRowNavigationOptionStopDisableRow);
        }
    }
    else if ([rowDescriptor.tag isEqualToString:kAccessoryViewRowNavigationStopInlineRow]){
        if ([[rowDescriptor.value valueData] isEqual:@(YES)]){
            self.form.rowNavigationOptions = self.form.rowNavigationOptions | XLFormRowNavigationOptionStopInlineRow;
        }
        else{
            self.form.rowNavigationOptions = self.form.rowNavigationOptions & (~XLFormRowNavigationOptionStopInlineRow);
        }
    }
    else if ([rowDescriptor.tag isEqualToString:kAccessoryViewRowNavigationSkipCanNotBecomeFirstResponderRow]){
        if ([[rowDescriptor.value valueData] isEqual:@(YES)]){
            self.form.rowNavigationOptions = self.form.rowNavigationOptions | XLFormRowNavigationOptionSkipCanNotBecomeFirstResponderRow;
        }
        else{
            self.form.rowNavigationOptions = self.form.rowNavigationOptions & (~XLFormRowNavigationOptionSkipCanNotBecomeFirstResponderRow);
        }
    }
}

-(UIView *)inputAccessoryViewForRowDescriptor:(XLFormRowDescriptor *)rowDescriptor
{
    if ([[[self.form formRowWithTag:kAccessoryViewRowNavigationShowAccessoryView].value valueData] isEqual:@NO]){
        return nil;
    }
    return [super inputAccessoryViewForRowDescriptor:rowDescriptor];
}

@end

File: Examples/Swift/SwiftExample/AccessoryViews/AccessoryViewFormViewController.swift
//
//  AccessoryViewFormViewController.swift
//  XLForm ( https://github.com/xmartlabs/XLForm )
//
//  Copyright (c) 2014-2015 Xmartlabs ( http://xmartlabs.com )
//
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.



class AccessoryViewFormViewController : XLFormViewController {


    fileprivate struct Tags {
        static let AccessoryViewRowNavigationEnabled = "RowNavigationEnabled"
        static let AccessoryViewRowNavigationShowAccessoryView     = "RowNavigationShowAccessoryView"
        static let AccessoryViewRowNavigationStopDisableRow        = "rowNavigationStopDisableRow"
        static let AccessoryViewRowNavigationSkipCanNotBecomeFirstResponderRow = "rowNavigationSkipCanNotBecomeFirstResponderRow"
        static let AccessoryViewRowNavigationStopInlineRow = "rowNavigationStopInlineRow"
        static let AccessoryViewName = "name"
        static let AccessoryViewEmail = "email"
        static let AccessoryViewTwitter = "twitter"
        static let AccessoryViewUrl = "url"
        static let AccessoryViewDate = "date"
        static let AccessoryViewTextView = "textView"
        static let AccessoryViewCheck = "check"
        static let AccessoryViewNotes = "notes"
    }

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        initializeForm()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        initializeForm()
    }

    func initializeForm() {
        
        let form : XLFormDescriptor
        var section : XLFormSectionDescriptor
        var row : XLFormRowDescriptor
        
        form = XLFormDescriptor(title: "Accessory View")
        form.rowNavigationOptions = .enabled
    
        // Configuration section
        section = XLFormSectionDescriptor()
        section.title = "Row Navigation Settings"
        section.footerTitle = "Changing the Settings values you will navigate differently"
        form.addFormSection(section)
    
        // RowNavigationEnabled
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewRowNavigationEnabled, rowType: XLFormRowDescriptorTypeBooleanSwitch, title: "Row Navigation Enabled?")
        row.value = true
        section.addFormRow(row)
    
        // RowNavigationShowAccessoryView
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewRowNavigationShowAccessoryView, rowType:XLFormRowDescriptorTypeBooleanCheck, title:"Show input accessory row?")
        row.value = form.rowNavigationOptions.contains(.enabled)
        row.hidden = "$\(Tags.AccessoryViewRowNavigationEnabled) == 0"
        section.addFormRow(row)

        // RowNavigationStopDisableRow
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewRowNavigationStopDisableRow, rowType: XLFormRowDescriptorTypeBooleanCheck, title:"Stop when reach disabled row?")
        row.value = form.rowNavigationOptions.contains(.stopDisableRow)
        row.hidden = "$\(Tags.AccessoryViewRowNavigationEnabled) == 0"
        section.addFormRow(row)
    
        // RowNavigationStopInlineRow
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewRowNavigationStopInlineRow, rowType: XLFormRowDescriptorTypeBooleanCheck, title: "Stop when reach inline row?")
        row.value = form.rowNavigationOptions.contains(.stopInlineRow)
        row.hidden = "$\(Tags.AccessoryViewRowNavigationEnabled) == 0"
        section.addFormRow(row)
        
        // RowNavigationSkipCanNotBecomeFirstResponderRow
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewRowNavigationSkipCanNotBecomeFirstResponderRow, rowType:XLFormRowDescriptorTypeBooleanCheck, title:"Skip Can Not Become First Responder Row?")
        row.value = form.rowNavigationOptions.contains(XLFormRowNavigationOptions.skipCanNotBecomeFirstResponderRow)
        row.hidden = "$\(Tags.AccessoryViewRowNavigationEnabled) == 0"
        section.addFormRow(row)
        
        section = XLFormSectionDescriptor()
        form.addFormSection(section)
        
        // Name
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewName, rowType: XLFormRowDescriptorTypeText, title: "Name")
        row.isRequired = true
        section.addFormRow(row)
        
        // Email
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewEmail, rowType: XLFormRowDescriptorTypeEmail, title: "Email")
        // validate the email
        row.addValidator(XLFormValidator.email())
        section.addFormRow(row)
        
        // Twitter
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewTwitter, rowType: XLFormRowDescriptorTypeTwitter, title: "Twitter")
        row.disabled = NSNumber(value: true as Bool)
        row.value = "@no_editable"
        section.addFormRow(row)
    
        // Url
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewUrl, rowType: XLFormRowDescriptorTypeURL, title: "Url")
        section.addFormRow(row)
        
    
        // Date
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewDate, rowType:XLFormRowDescriptorTypeDateInline, title:"Date Inline")
        row.value = Date()
        section.addFormRow(row)

        
        
        section = XLFormSectionDescriptor()
        form.addFormSection(section)
        
        
        
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewTextView, rowType:XLFormRowDescriptorTypeTextView)
        row.cellConfigAtConfigure["textView.placeholder"] = "TEXT VIEW EXAMPLE"
        section.addFormRow(row)
        
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewCheck, rowType:XLFormRowDescriptorTypeBooleanCheck, title:"Check")
        section.addFormRow(row)

        section = XLFormSectionDescriptor()
        form.addFormSection(section)
        
        row = XLFormRowDescriptor(tag: Tags.AccessoryViewNotes, rowType:XLFormRowDescriptorTypeTextView, title:"Notes")
        section.addFormRow(row)
    
        self.form = form
    }

    
    override func inputAccessoryView(for rowDescriptor: XLFormRowDescriptor!) -> UIView! {
        if (form.formRow(withTag: Tags.AccessoryViewRowNavigationShowAccessoryView)!.value! as AnyObject).boolValue == false {
            return nil
        }
        return super.inputAccessoryView(for: rowDescriptor)
    }
    
    
//MARK: XLFormDescriptorDelegate
    
    override func formRowDescriptorValueHasChanged(_ formRow: XLFormRowDescriptor!, oldValue: Any!, newValue: Any!) {
        super.formRowDescriptorValueHasChanged(formRow, oldValue: oldValue, newValue: newValue)
        if formRow.tag == Tags.AccessoryViewRowNavigationStopDisableRow {
            if (formRow.value! as AnyObject).boolValue  == true {
                form.rowNavigationOptions = form.rowNavigationOptions.union(.stopDisableRow)
            }
            else{
                form.rowNavigationOptions = form.rowNavigationOptions.subtracting(.stopDisableRow)
            }
        }
        else if formRow.tag == Tags.AccessoryViewRowNavigationStopInlineRow {
            if (formRow.value! as AnyObject).boolValue  == true {
                form.rowNavigationOptions = form.rowNavigationOptions.union(.stopInlineRow)
            }
            else{
                var options = form.rowNavigationOptions
                options = options.subtracting(XLFormRowNavigationOptions.stopInlineRow)
                form.rowNavigationOptions = options
            }
        }
        else if formRow.tag == Tags.AccessoryViewRowNavigationSkipCanNotBecomeFirstResponderRow {
            if (formRow.value! as AnyObject).boolValue  == true {
                form.rowNavigationOptions = form.rowNavigationOptions.union(.skipCanNotBecomeFirstResponderRow)
            }
            else{
                var options = form.rowNavigationOptions
                options = options.subtracting(.skipCanNotBecomeFirstResponderRow)
                form.rowNavigationOptions = options
            }
        }
        
    }

}
